Implementation of Deferred FIPS Self-Tests

simo5 opened the pull request:
Start implementing the `Deferred FIPS Self-Tests` design

This PR goes beyond what was initially described in the design and completely unifies the self-test infrastructure as well to make it easier to deal for all cases.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [x] tests are added or updated

slontis commented on providers/fips/self_test_data.c:
```diff
@@ -201,33 +73,51 @@ static const unsigned char sha3_256_digest[] = {
     },
+    {
+         OSSL_SELF_TEST_DESC_MD_SHA2,
+         "SHA256",
```
This entry would not be ondemand..
So just SELF_TEST_FLAG_DEFERRED

slontis commented on providers/implementations/ciphers/cipher_aes_ccm.c:
```diff
@@ -24,8 +29,7 @@ static void *aes_ccm_newctx(void *provctx, size_t keybits)
 
-    if (!ossl_prov_is_running())
-        return NULL;
+    DIGEST_PROV_CHECK(provctx, aes);
```
If you are going to use a macro please give it the right name for a cipher.

simo5 replied:
yeah, I am not yet happy with this macro, but naming it right is kinda necessary.

slontis commented on providers/fips/self_test_data.c:
```diff
@@ -3299,6 +3210,68 @@ static const ST_KAT_SIGN st_kat_sign_tests[] = {
+#endif
+
+#ifndef OPENSSL_NO_SLH_DSA
+/* mark one as dependent of the other, so both are executed */
```
"dependent on" OR "a dependent of"

slontis commented on providers/fips/self_test_data.c:
```diff
@@ -3299,6 +3210,68 @@ static const ST_KAT_SIGN st_kat_sign_tests[] = {
 };
+int st_kat_sign_tests_size = OSSL_NELEM(st_kat_sign_tests);
+
+FIPS_DEFERRED_TEST *sig_sha2_also_satisfies[] = {
```
Should this be sig_sha256_also_satisfies[]

slontis commented on providers/fips/self_test_data.c:
```diff
@@ -3678,10 +3651,11 @@ static const ST_KAT_KEM st_kat_kem_tests[] = {
         ml_kem_512_reject_secret    /* No length because same as _secret's */
     },
 };
+int st_kat_kem_tests_size = OSSL_NELEM(st_kat_kem_tests);
```
const ?

slontis commented on providers/fips/self_test_data.c:
```diff
@@ -3299,6 +3210,68 @@ static const ST_KAT_SIGN st_kat_sign_tests[] = {
+    NULL
+};
+FIPS_DEFERRED_TEST slh_sig_deferred_test = {
+    "SLH-DSA-SHAKE-128f",
```
SLH-DSA-SHAKE-128f depends on SHAKE256

slontis commented:
Please read https://openssl-library.org/policies/technical/coding-style/

simo5 commented on providers/implementations/include/prov/ciphercommon.h:
```diff
@@ -191,6 +191,18 @@ const OSSL_DISPATCH ossl_##alg##kbits##lcmode##_functions[] = {                \
     OSSL_DISPATCH_END                                                          \
 };
 
+# if defined(FIPS_MODULE) && defined(CIPHER_IS_FIPS)
```
I would like to avoid having to parse large arrays of tests, but I can think of some way to do this once all tests have been converted, by using an enum as an index.

simo5 replied:
enums have been implemented

simo5 commented on providers/fips/self_test_data.c:
```diff
@@ -3678,10 +3651,11 @@ static const ST_KAT_KEM st_kat_kem_tests[] = {
 
 #if !defined(OPENSSL_NO_ML_KEM) || !defined(OPENSSL_NO_ML_DSA) || !defined(OPENSSL_NO_SLH_DSA)
-static const ST_KAT_ASYM_KEYGEN st_kat_asym_keygen_tests[] = {
+const ST_KAT_ASYM_KEYGEN st_kat_asym_keygen_tests[] = {
```
they are exposed only to self_test_kats.c, which really need access to the structures.

simo5 commented on providers/fips/self_test_data.c:
```diff
@@ -3299,6 +3210,68 @@ static const ST_KAT_SIGN st_kat_sign_tests[] = {
+#ifndef OPENSSL_NO_SLH_DSA
+/* mark one as dependent of the other, so both are executed */
+FIPS_DEFERRED_TEST slh_dsa_sha2_sig_deferred_test = {
+    "SLH-DSA-SHA2-128f",
```
I would add an also_satisfies for HMAC as well, but we do not have a pure HMAC test yet ... will add one later I guess...

paulidale commented on providers/fips/self_test_data.c:
```diff
@@ -3299,6 +3210,68 @@ static const ST_KAT_SIGN st_kat_sign_tests[] = {
+    "RSA",
+    FIPS_DEFERRED_KAT_SIGNATURE,
+    FIPS_DEFERRED_TEST_INIT,
+    sig_sha2_also_satisfies,
```
Could called algorithms set their tested state when they are first called?  I.e. do it eagerly.

If such a test later failed, the overall test would and the module would enter the error state and the one errantly flaged as passed testing wouldn't be usable.

slontis commented on providers/fips/self_test.c:
```diff
@@ -257,7 +199,7 @@ static int verify_integrity(OSSL_CORE_BIO *bio, OSSL_FUNC_BIO_read_ex_fn read_ex
     OSSL_PARAM params[2], *p = params;
 
-    if (!integrity_self_test(ev, libctx))
+    if (!SELF_TEST_kats_single(ev, libctx, ST_ID_MAC_HMAC))
```
Could you mark SHA2 as being done at this point?

slontis commented on providers/fips/self_test_data.c:
```diff
@@ -3692,79 +3239,643 @@ static const ST_KAT_PARAM ml_kem_key[] = {
+        "HMAC",
+        OSSL_SELF_TEST_DESC_INTEGRITY_HMAC,
+        SELF_TEST_KAT_MAC,
+        SELF_TEST_DEFERRED,
```
HMAC is not deferred

slontis replied:
You probably need a separate state for this one, since it is not onload either, its a integrity phase check only.

slontis replied:
When we do 'on demand' tests i would like it to just run the tests it was running before you added deferred testing..
i.e. NOT SHA2 for example. The same holds if the user decides not to run in deferred mode.

slontis commented on providers/fips/self_test_data.c:
```diff
@@ -3692,79 +3239,643 @@ static const ST_KAT_PARAM ml_kem_key[] = {
-static const ST_KAT_ASYM_KEYGEN st_kat_asym_keygen_tests[] = {
-# if !defined(OPENSSL_NO_ML_KEM)
+ST_DEFINITION st_all_tests[ST_ID_MAX] = {
+    {
```
This data table and the associated enum should be generated so that there is no way that they can be out of synch.

slontis replied:
Generating a file does not make it unreadable.

slontis replied:
Its a maintenance problem currently for anyone other than the person who added this PR.
Having a large enum table that has to exactly match the order in fairly large table is fragile. This should be documented in both the enum and the table if you are going to do this.
I got this wrong just about straight away when I tried removing the DRBG fields (but didnt remove the enum id's).

slontis replied:
Please add a comment to the enum

slontis commented on apps/fipsinstall.c:
```diff
@@ -197,6 +199,7 @@ typedef struct {
     unsigned int x942kdf_key_check : 1;
     unsigned int pbkdf2_lower_bound_check : 1;
     unsigned int ecdh_cofactor_check : 1;
+    unsigned int no_defer_tests : 1;
```
We probably now need another option here that allows the self tests to run on demand now.
Since the 'defer_tests' no longer runs all tests, it may be nice for the user to be able to check that the tests will work when they do run.

simo5 replied:
Uhmm I thought the change wa also causing the tests to be run immediately as I think fipsinstall uses its own configuration. If it doesn't I think I would still tie the two actions together, as I can't think of a case where it wuld make sense to want to force all tests at startup but not do it for fipsinstall itself.

slontis reviewed (commented):
This needs formal tests to check that it functions as expected.
You could do something similar to provider_status_test which sets up self tests callbacks..
Then test against many combinations...
e.g.
on demand followed by keygen expected results.
keygen expected results.

slontis reviewed (commented):
Can you see any reason why we cant defer the DRBG self tests now?
I tried deferring them and it seems to work..
I also then tried several combinations of running normal crypto before and after with deferred and on demand and it seemed to work..
Was there a reason why this doesnt work?

simo5 commented:
> Now that you use ids you no longer need self_test_data.h, and self_test_data.c can just become a .inc again?

No please, .inc is a "nice" hack, but remains a hack, I prefer regular source code whenever possible, and I see no reason why this code should not be just regular code.

slontis commented:
> > Now that you use ids you no longer need self_test_data.h, and self_test_data.c can just become a .inc again?
> 
> No please, .inc is a "nice" hack, but remains a hack, I prefer regular source code whenever possible, and I see no reason why this code should not be just regular code.

There is nothing external about this file, and it is just a big blob of data used by one file..

slontis reviewed (commented):
Just to reiterate, if the user choses to not use deferred mode then I would expect minimal tests to be run when composite algorithms are used.  (i.e. HMAC not run multiple times, and SHA2 not to be run). If this needs to be flagged manually in the table entries that is ok

slontis commented:
I like how you have dynamically figured out the dependancies.

slontis commented:
> @slontis so the AES ECB test now fails because AES_ECB is used by the DRBG test which I now reorganized to happen first and therefore it is marked as implicitly tested.Due to that the test is actually not run, and therefore the callback is not called...
> 
> Ideas?

Are you talking about the fipsinstall corruption test in 03-test_fipsinstall.t?
Those tests were put in there to test the corruption paths (I believe this is required as part of the NIST (VE) Vendor Evidence documentation).
This looks like a case where we would want it to run the dependancies instead of just marking them as done.
So I would propose this as some 'test' option that we add that gets set via the fipsinstall that is passed through to the fips provider (the same way we pass other options by setting them into the config). This option would not normally be used.

slontis commented:
The other thing to note is that the Symmetric Ciphers CAST requirements are:
1) Encryption Test using a Forward Cipher (This covers all AES ciphers for Encryption)
2) Decryption Test using a Forward Cipher (This covers modes that are not ECB,CBC,XTS,KW,KWP)
3) An Autheticated Encryption Test (e.g. AES-GCM)
4) An inverse cipher Test (This covers decryption for ECB,CBC,XTS,KW,KWP)

To cover 1,2,3 we do An AES-GCM encrypt/decrypt
To cover 4 we do AES-ECB decryption.

So
HASH-DRBG only covers SHA256
HMAC-DRBG only covers HMAC + SHA256
CTR-DRBG only covers AES-CTR and AES-ECB encryption which is 1)
This means that 2) 3) and 4) are not covered by CTR-DRBG.

simo5 commented:
> The other thing to note is that the Symmetric Ciphers CAST requirements are:
> 
>     1. Encryption Test using a Forward Cipher (This covers all AES ciphers for Encryption)
> 
>     2. Decryption Test using a Forward Cipher (This covers modes that are not ECB,CBC,XTS,KW,KWP)
> 
>     3. An Authenticated Encryption Test (e.g. AES-GCM)
> 
>     4. An inverse cipher Test (This covers decryption for ECB,CBC,XTS,KW,KWP)
> 
> 
> To cover 1,2,3 we do An AES-GCM encrypt/decrypt To cover 4 we do AES-ECB decryption.
> 
> So HASH-DRBG only covers SHA256 HMAC-DRBG only covers HMAC + SHA256 CTR-DRBG only covers AES-CTR and AES-ECB encryption which is 1) This means that 2) 3) and 4) are not covered by CTR-DRBG.

I can simply make AES_ECB test an explicit dependency of CTR_DRBG, this means 4 would be explicitly tested before the CTR_DRBG test is run. (This is exactly what the depends_on feature was built for, a test that does not fully exercises the test requirements on its own).

simo5 commented:
So adding a depends on CTR_DRBG solved the AES issue as expected, but now we get failures for SHA2 and SHA3 corruption checks, I think I will deal with the whack-a-mole game in the next few days ... meanwhile I would love some more review on the code architecture and the other choices made.

beldmit reviewed (dismissed):
I don't see any obvious errors with this PR. LGTM.

t-j-h commented:
> @t-j-h Is it possible to get a OpenSSL resource to look at this PR?

Thanks for taking a look so far - I'll get someone to take a pass through it - as we definitely want this capability into the release. Labs generally review the code for these and don't explicitly test every combination - relying on code review. I'd expect we will get this PR in first and then see if we need to tackle a comprehensive testing approach for all combinations separately. Does that make sense to you?

slontis commented on providers/fips/fipsprov.c:
```diff
@@ -1318,26 +1322,18 @@ static int FIPS_kat_deferred(OSSL_LIB_CTX *libctx, FIPS_DEFERRED_TEST *test)
-
-        if (ev)
-            OSSL_SELF_TEST_free(ev);
+        if (st)
```
if not required.

slontis commented on providers/fips/self_test_kats.c:
```diff
@@ -1209,235 +1089,174 @@ static int setup_main_random(OSSL_LIB_CTX *libctx)
+        break;
     }
+    if (ret)
+        st_all_tests[id].state = SELF_TEST_STATE_PASSED;
```
should this check the state before setting it to passed?

slontis commented on providers/implementations/kdfs/pbkdf2.c:
```diff
@@ -228,7 +237,13 @@ static int pbkdf2_lower_bound_check_passed(int saltlen, uint64_t iter,
+#ifdef FIPS_MODULE
+    /* Modify this check during self-test. See FIPS 140-3 IG 10.3.A.8 */
+    if (ossl_self_test_in_progress(ST_ID_KDF_PBKDF2)) {
+        min_iter = 2;
```
Put the magic number at the top of the file as a define.

slontis reviewed (commented):
By manual testing I can see that AES-128-CTR encrypt runs just the inverse cipher test, which is not sufficient. I can't emphasize enough that this needs to be tested thoroughly before this is allowed.

slontis commented:
> Thanks for taking a look so far - I'll get someone to take a pass through it - as we definitely want this capability into the release. Labs generally review the code for these and don't explicitly test every combination - relying on code review. I'd expect we will get this PR in first and then see if we need to tackle a comprehensive testing approach for all combinations separately. Does that make sense to you?

@t-j-h So you dont mind if there are some holes on the first merge, I assume there will not be another FIPS release until some later 4.X release?

simo5 commented on providers/fips/self_test_data.h:
```diff
@@ -0,0 +1,45 @@
+ * they are tested as part of a higher level algorithm (such as HMAC).
+ */
+
+extern const ST_DEFINITION st_kat_digest_tests[];
```
good catch, fixing.

simo5 commented:
> By manual testing I can see that AES-128-CTR encrypt runs just the inverse cipher test, which is not sufficient. I can't emphasize enough that this needs to be tested thoroughly before this is allowed.

Which other test should it run ?

simo5 commented:
@slontis I think I have addressed all your issues except two on which I need clarification. Once I know the answers to sshkdf and AES-128-CTR testing I will fix them and do a final rebase and squash-in of the fixups

slontis commented on providers/implementations/kdfs/sshkdf.c:
```diff
@@ -61,6 +62,16 @@ static void *kdf_sshkdf_new(void *provctx)
 
+#ifdef FIPS_MODULE
+    /*
+     * According to FIPS 140-3 10.3.A Note18: SSH KDF is not required, since
```
yes - that is what seems to happen when I removed the line locally (when doing HMAC-SHA1)

slontis commented:
> > By manual testing I can see that AES-128-CTR encrypt runs just the inverse cipher test, which is not sufficient. I can't emphasize enough that this needs to be tested thoroughly before this is allowed.
> 
> Which other test should it run ?

See the discussion in https://github.com/openssl/openssl/pull/29222#issuecomment-3675316216.
We currently do AES_GCM to cover the forward cipher and the Authenticated case (See FIPS 140-3 IG related to requirements for CAST for ciphers, and which ones relate to forward and inverse).
So we either use AES-GCM here or have a forward cipher test..

This will fix this one - there could be more..

simo5 commented:
@slontis so what I did was to add AES-GCM tests as a dependency for AES-ECB tests (which are the tests normally run for most "AES" ciphers.

Additionally I noticed that CTR DRBG would cause AES to be marked as passed without running the dependencies, so I also added code to run explicit dependencies for tests that are marked as implicitly tested.

This should cover it all correctly now.

t8m commented on test/recipes/03-test_fipsinstall.t:
```diff
@@ -263,6 +263,7 @@ SKIP: {
                  '-module', $infile,
                  '-provider_name', 'fips', '-mac_name', 'HMAC',
                  '-macopt', 'digest:SHA256', '-macopt', "hexkey:$fipskey",
+                 '-defer_tests', '0',
```
What is this `0`? `-defer_tests` should not have any arguments, does it?

simo5 replied:
IIRC:
'1' defers tests
'0' does not

simo5 replied:
(the default is set to 1, but happy to change it for fipsinstall and drop the argument if you prefer it that way)

simo5 commented:
I think our style checks are still broken, the style errors reported in this PR are really issues in the original PR that was just merged yesterday and I rebased onto ...

paulidale commented:
> Reviewing the deferred test execution of fipsinstall (prompted by Tomas' comment) I realized I had promised to always run all tests in fipsinstall by default, as it makes little sense to defer test in this app.

With the change to not allow on install self tests, fipsinstall doesn't have to run any tests.
It's sensible for it to test everything still but not required.

t8m commented on test/recipes/00-prep_fipsmodule_cnf.t:
```diff
@@ -30,7 +30,7 @@ my $fipsmoduleconf = bldtop_file('test', 'fipsmodule.cnf');
 
 # Create the $fipsmoduleconf file
-ok(run(app(['openssl', 'fipsinstall', '-pedantic',
+ok(run(app(['openssl', 'fipsinstall', '-pedantic', '-defer_tests',
```
So this basically means we always run our `make test` tests with the deferred self test enabled. Which is fine but we should perhaps also have a test case that overrides it and runs without that?

simo5 replied:
I could simply drop this and always run self-tests, the deferred tests code is tests by all other openssl command line invocations anyway

simo5 commented:
@openssl/committers one more ping, keeping this PR up to date and rebasing it constantly is a bit tedious and error prone. I would be very grateful if I can get the two reviews needed to merge it in. We should have it in 4.0 pre long enough to work out if there are any unexpected side-effects to deal with.

slontis commented:
@beldmit I will try to review this again soon. I need to run locally to test some things related to the how the defer options now work.

slontis commented on providers/fips/self_test.c:
```diff
@@ -383,11 +325,26 @@ int SELF_TEST_post(SELF_TEST_POST_PARAMS *st, int on_demand_test)
+        /* ensure all states are set so no test is passd implicitly,
+         * this forces each test to always run regardless of order */
+        for (int i = 0; i < ST_ID_MAX; i++) {
+            st_all_tests[i].state = SELF_TEST_STATE_NO_IMPLICIT;
```
Why does the on demand above set to STATE_INT and this one does STATE_NO_IMPLICIT?

slontis commented on providers/fips/self_test.h:
```diff
@@ -31,8 +35,143 @@ typedef struct self_test_post_params_st {
+    SELF_TEST_STATE_IN_PROGRESS,
+    SELF_TEST_STATE_PASSED,
+    SELF_TEST_STATE_FAILED,
+    SELF_TEST_STATE_IMPLICIT,
```
Maybe describe these last 2 in a comment here?

slontis commented:
> What is the value for do-not-defer-tests in the fispmodule.cnf included for these tests?

It is 0 as described above

slontis commented on doc/man7/OSSL_PROVIDER-FIPS.pod:
```diff
@@ -327,10 +327,11 @@ Uses HMAC SHA256 on the module file to validate that the module has not been
 file during installation.
 
-=item "KAT_Integrity" (B<OSSL_SELF_TEST_TYPE_KAT_INTEGRITY>)
+=item "KAT_Mac" (B<OSSL_SELF_TEST_TYPE_KAT_MAC>)
```
Ok I was just thinking it might get lost with time.

simo5 commented on providers/implementations/kdfs/sshkdf.c:
```diff
@@ -61,6 +62,16 @@ static void *kdf_sshkdf_new(void *provctx)
     if (!ossl_prov_is_running())
         return NULL;
 
+#ifdef FIPS_MODULE
```
I figured I'd make clear where a test would go if we had to have one (or if in future NIST will require one) :-)

simo5 commented on include/internal/fips.h:
```diff
@@ -11,46 +11,125 @@
+    ST_ID_SIG_DET_ECDSA_SHA256,
+#endif
+#ifndef OPENSSL_NO_EC2M
+    ST_ID_SIG_NO_E2CM_ECDSA_SHA256,
```
But then no code points at this test, so I am probably missing an invocation somewhere

slontis commented:
Thanks for doing the work in this PR. (I would have done it eventually)

For future PR's could you please use fix-up commits a bit more when it gets to the review stage,
it makes it quite a bit easier to track what has changed (Otherwise you end up having to re-review all files which takes a while when there are lots of files). Did you collapse them maybe when rebasing?
Your last commit is this
https://github.com/openssl/openssl/pull/29222/changes/072ae83a8e8b919e1779d5d0585872bb6fbf2761 which just seems to be formatting related.
(The compare tab doesnt work well in github since in includes whatever was merged from master for the rebase).

slontis commented:
The tests failing above need to be addressed.

simo5 commented:
> The failing CI tests need to be fixed. These tests run both forwards and backward compatible tests between different fips providers and versions of OpenSSL. So it does..
> 
>     1. your PR but copy the fips provider from different versions of OpenSSL
>        We might have to enable this label on another PR, as I am not sure the SSLtest failures are related to this PR.
> 
>     2. copy the fips provider from your PR to different versions of OpenSSL
>        This is probably a result of changing the defer tests to run by default, for backwards compat it should probably not defer.
> 
> 
> These tests are only done if you add the 'extended test' label which is why you did not see it before.

Sounds good, should we attempt to not-defer by default, as a way to handle bacward compat?
This will cause some churn again in the code, and a reversal of the knobs for fipsinstall and friends, but would be the more conservative way to introduce this change I think.

slontis commented:
I think the SSL one is caused by this (https://github.com/openssl/openssl/pull/29962) - I was going to add the label to the latest PR, but that seemed to be the PR that tried to fix it :) Easiest way to test in future would be to create a dummy PR with some small change such as a doc change and add the 'extended tests' label.

slontis commented:
It seems like the best approach.

slontis commented:
Should be done with the doc updates.

slontis commented on doc/man1/openssl-fipsinstall.pod.in:
```diff
@@ -396,6 +397,12 @@ data that is included by the base C<parent_config> configuration file.
 See L<config(5)> for further information on how to set up a provider section.
 All other options are ignored if '-config' is used.
 
+=item B<-defer_tests>
```
Please update the history section.

slontis commented on doc/man1/openssl-fipsinstall.pod.in:
```diff
@@ -494,6 +494,10 @@ B<-x963kdf_key_check>,
 B<-x942kdf_key_check>,
 B<-ecdh_cofactor_check>
 
+The following options was added in OpenSSL 4.0:
```
NIT option.
